CCIP-4165 additional e2e integration tests

jhweintraub opened the pull request:
Additional End-to-End integration tests for CCIP involving token transfers paying in different feeTokens

jhweintraub committed:
additional e2e integration test first version

makramkd commented on integration-tests/smoke/ccip_test.go:
```diff
@@ -276,3 +277,460 @@ func TestTokenTransfer(t *testing.T) {
 }
+
+func Test_PricingForTokenTransfers(t *testing.T) {
+	// Deplot the environment
```
```suggestion
	// Deploy the environment
```

makramkd commented on integration-tests/smoke/ccip_test.go:
```diff
@@ -276,3 +277,460 @@ func TestTokenTransfer(t *testing.T) {
+		new(big.Int).Mul(twoCoins, big.NewInt(10)),
+	)
+
+	if err != nil {
+		fmt.Printf("Error minting token: %v\n", err)
+	}
```
Can use `require.NoError` here instead of this, since the test cannot move forward if this fails

jhweintraub replied:
Remnant from Debugging. Good Catch thanks!

makramkd commented on integration-tests/smoke/ccip_test.go:
```diff
@@ -276,3 +277,460 @@ func TestTokenTransfer(t *testing.T) {
+		fmt.Printf("Error minting token: %v\n", err)
+	}
+
+	// Confirm that mint tx
+	require.NoError(t, err)
```
Seems like you do that here, the above error check is redundant

makramkd replied:
Btw, can move this `NoError` call closer to the call site of `Mint`

jhweintraub committed:
attempt fix linter

makramkd commented on integration-tests/smoke/ccip_test.go:
```diff
@@ -276,3 +277,460 @@ func TestTokenTransfer(t *testing.T) {
+	require.NoError(t, err)
+
+	_, err = e.Chains[tenv.HomeChainSel].Confirm(tx)
+	require.NoError(t, err)
```
To aid in readability, can we have all of this setup in a separate function, with a docstring describing what is being done? e.g 

```golang
// setupPricingTest deploys transferable tokens on the source and dest,
// mints tokens on both, ....
func setupPricingTest(...) {
// the above code, from lines 352 to 399
}
```

jhweintraub replied:
Sure. I'll see what I can do.

makramkd commented on integration-tests/smoke/ccip_test.go:
```diff
@@ -276,3 +277,460 @@ func TestTokenTransfer(t *testing.T) {
+		}
+
+		// Get the fee Token Balance Before
+		feeTokenBalanceBefore, _ := linkToken.BalanceOf(nil, e.Chains[tenv.HomeChainSel].DeployerKey.From)
```
Should check error here

makramkd commented on integration-tests/smoke/ccip_test.go:
```diff
@@ -276,3 +277,460 @@ func TestTokenTransfer(t *testing.T) {
+		expectedSeqNum[dest] = seqNum
+
+		// Check the fee token balance after the request and ensure fee tokens were spent
+		feeTokenBalanceAfter, _ := linkToken.BalanceOf(nil, e.Chains[tenv.HomeChainSel].DeployerKey.From)
```
Check error

makramkd commented on integration-tests/smoke/ccip_test.go:
```diff
@@ -276,3 +277,460 @@ func TestTokenTransfer(t *testing.T) {
+	_, err = e.Chains[tenv.HomeChainSel].Confirm(tx)
+	require.NoError(t, err)
+
+	t.Run("Send Token Pay with Link token home chain -> remote", func(t *testing.T) {
```
q: are these test cases runnable separately or do they depend on each other?

jhweintraub replied:
They can be run independently, the ticket request asked that they be put in this way and run sequentially under the umbrella of a single test.

makramkd reviewed (commented):
Awesome! Are there any commonalities from each test case that can be pulled out into funcs? See e.g `messagingTestCase` in ccip_messaging_test.go

jhweintraub committed:
test cleanup and linting

jhweintraub committed:
Merge branch 'develop' into CCIP-4165-e-2-e-tests-implement-pricing-for-tokens-transfers-tests

jhweintraub committed:
remove duplicate import

jhweintraub committed:
remove additional message data from ccipSendRequest

jhweintraub committed:
Merge branch 'develop' into CCIP-4165-e-2-e-tests-implement-pricing-for-tokens-transfers-tests

jhweintraub committed:
attempt lint fix and comment

jhweintraub committed:
Merge branch 'develop' into CCIP-4165-e-2-e-tests-implement-pricing-for-tokens-transfers-tests

jhweintraub committed:
add new e2e messaging tests

jhweintraub committed:
checkpoint work

jhweintraub committed:
Merge branch 'develop' into CCIP-4165-e-2-e-tests-implement-pricing-for-tokens-transfers-tests

jhweintraub committed:
checkpointing

jhweintraub committed:
Merge branch 'develop' into CCIP-4165-e-2-e-tests-implement-pricing-for-tokens-transfers-tests

jhweintraub committed:
test passing

jhweintraub committed:
Merge branch 'develop' into CCIP-4165-e-2-e-tests-implement-pricing-for-tokens-transfers-tests

jhweintraub committed:
fix smoke tests after they broke in the last merge from develop

jhweintraub committed:
message and programmable token transfer e2e tests

jhweintraub committed:
bump smoke test timeout parameter

jhweintraub committed:
bump time again

jhweintraub committed:
Merge branch 'develop' into CCIP-4165-e-2-e-tests-implement-pricing-for-tokens-transfers-tests

b-gopalswami commented on integration-tests/smoke/ccip/ccip_test.go:
```diff
@@ -206,3 +208,797 @@ func TestTokenTransfer(t *testing.T) {
+		require.NoError(t, err)
+		require.GreaterOrEqual(t, srctokenBalance.Int64(), twoCoins.Int64())
+
+		// Assign Src to the Home Chain Selector and destChain to the Feed Chain Selector
```
Is this comment relevant here?

b-gopalswami commented on integration-tests/smoke/ccip/ccip_test.go:
```diff
@@ -206,3 +208,797 @@ func TestTokenTransfer(t *testing.T) {
+		balance, err := dstToken.BalanceOf(nil, state.Chains[dest].Receiver.Address())
+		require.NoError(t, err)
+
+		// The balance should be 4 since we've already sent 2 tokens over this lane direction in the
+		// first part of the test, so balance should already be two
```
It's better to avoid sequential testing as this will cost the execution time. I would advice to see how we can make these test runs in parallel.
Few options are like:
1. Using multiple wallet address
2. Creating it as individual test which has its own setup and runner to run in parallel.

jhweintraub replied:
running in parallel with multiple wallets makes the most sense as the original ticket specified under the same e2e test.

b-gopalswami commented on integration-tests/smoke/ccip/ccip_test.go:
```diff
@@ -206,3 +208,797 @@ func TestTokenTransfer(t *testing.T) {
+		delete(expectedSeqNum, sourceDestPair)
+	})
+
+	t.Run("Send Token Pay with native remote chain -> home", func(t *testing.T) {
```
If the difference in this test is only about fee token, then we can pass that as an argument and simplify the test.

b-gopalswami commented on integration-tests/smoke/ccip/ccip_test.go:
```diff
@@ -206,3 +208,797 @@ func TestTokenTransfer(t *testing.T) {
+		}},
+	}
+
+	t.Run("Send message Pay with Link token home chain -> remote", func(t *testing.T) {
```
As our intention is to test fee token, I think we can avoid testing this is in different lanes. We can test it in the same lane with different token by that we can simplify the test more.

jhweintraub replied:
The original ticket specified that each test scenario should flip the lane direction

b-gopalswami replied:
I feel it is not needed. @makramkd @mateusz-sekara What do you guys think about this?

b-gopalswami commented on integration-tests/smoke/ccip/ccip_test.go:
```diff
@@ -206,3 +208,797 @@ func TestTokenTransfer(t *testing.T) {
+		delete(expectedSeqNum, sourceDestPair)
+	})
+
+	t.Run("Send message pay with wrapped native home chain -> remote", func(t *testing.T) {
```
We can do something like this:
1. Setup one lane say home chain -> remote
2. Create test input struct with fee token type (link, native and wnative) and message type (Token, Data and Token+Data)
3. Loop the test and try to run in parallel using multiple wallet address. 

WDYT?

makramkd committed:
Merge branch 'develop' into CCIP-4165-e-2-e-tests-implement-pricing-for-tokens-transfers-tests

makramkd committed:
fix flakey assertion

use the fee token paid from the event rather than the call before the
request is made

makramkd committed:
refactor, use sim backend

makramkd committed:
fix workflow

makramkd committed:
fix lint
